### ISSUE TITLE
Doesn't work with Wayland

### ISSUE THREAD

**Issue by moceap** [grwlf/xkb-switch#64]
Can't detect current keyboard layout in Wayland session.

**Comment by grwlf** [comment]
Unfortunately, it is a strictly X11 app right now. Try to find another tool for Wayland.

**Comment by timsofteng** [comment]
@grwlf is it really complicated issue to add Wayland switching? E.g for sway. 
The interface is similar to xkb and could be invoke with sway-msg from vim.

**Comment by grwlf** [comment]
Originally I had no Wayland devices to think about it's support, but now I have a Pinephone which uses Wayland, so at least I could try now.

I think that it is not a good idea to integrate Wayland support to the originally-X11 application, but we may think about writing a separate tool with the same command-line interface.

Do you have any good examples of working with keyboard using this protocol?

**Comment by timsofteng** [comment]
@grwlf I use sway and layout can be changed with swaymsg externally.
It should not be hard to implement it.

**Comment by grwlf** [comment]
Afaik, X applications should run under Wayland's X compatibility layer called XWayland https://wayland.freedesktop.org/docs/html/ch05.html

**Comment by 1nikolas** [comment]
> Afaik, X applications should run under Wayland's X compatibility layer called XWayland https://wayland.freedesktop.org/docs/html/ch05.html Has anybody tried it with `xkb-switch`?

Yes it does run under xwayland but it seems xwayland doesn't sync the keyboard layouts with wayland.. It's like its own separate thing, I can set my layouts with setxkbmap and switch them/query them with xkb-switch but they don't affect the real keyboard layout in wayland

**Comment by FrolovRuslan1** [comment]
> Afaik, X applications should run under Wayland's X compatibility layer called XWayland https://wayland.freedesktop.org/docs/html/ch05.html Has anybody tried it with `xkb-switch`?

I can help develop `xkb-switch` under Wayland. Wayland support [libxkbcommon](https://xkbcommon.org/doc/current/index.html). [Here](https://github.com/FrolovRuslan1/wxkb_switch/tree/dev) is my test code where I can get list of layouts. BUT how I found out [here](https://github.com/xkbcommon/libxkbcommon/issues/409) libxkbcommon doesnt have tools to switch layouts. 
Here tests on Debian:
In Plasma Wayland:
1. 
```
-> % xkb-switch --list
us
```
2.
```
-> % ./wxkb_switch    
It prints keyboard layout only via xkbcommon
num: 1
English (US)
It prints keyboard layout via xkbcommon and wayland
num: 2
English (US)
Russian
```

In Plasma X11:
1.
```
-> % xkb-switch --list
us
ru
```
2.
```
-> % ./wxkb_switch
It prints keyboard layout only via xkbcommon
num: 1
English (US)
Failed to connect to Wayland display
```

**Comment by grwlf** [comment]
@FrolovRuslan1 I would be happy to accept the PR. Have some questions:

1. What do you think, can we write an abstraction which would run the existing code on X, and the wayland code you have prototyped if we are on wayland?  I think we can depart from making [this class](https://github.com/grwlf/xkb-switch/blob/master/src/XKeyboard.hpp) a C-structure with function pointers pointing to either X11 or Wayland implementations depending on the build configuration flag? (I am ok with moving towards pure-C implementation)
2. What changes do we need in the CMake?

**Comment by FrolovRuslan1** [comment]
> @FrolovRuslan1 I would be happy to accept the PR. Have some questions:
> 
>     1. What do you think, can we write an abstraction which would run the existing code on X, and the wayland code you have prototyped if we are on wayland?  I think we can depart from making [this class](https://github.com/grwlf/xkb-switch/blob/master/src/XKeyboard.hpp) a C-structure with function pointers pointing to either X11 or Wayland implementations depending on the build configuration flag? (I am ok with moving towards pure-C implementation)
> 
>     2. What changes do we need in the CMake?

I checked `XKeyboard.hpp` and yes, we can write an abstraction for Wayland. I think we can create Class automatically depends on environment variables. Or we can split the project on two. One for the X11 and one for Wayland because if we will be make project based on dependence of build configuration flag or runtime variable then the project growing up to unreadable size, moreover, I do not think that a new X11 analogue will appear under Linux in the near future.

**Comment by lolmaus** [comment]
Any progress on this one? :pray:

**Comment by FrolovRuslan1** [comment]
> Any progress on this one? 🙏

I hope that I get time at this summer for work at  xkb-switch and my openswitcher projects, I have lot of work and study

**Comment by FrolovRuslan1** [comment]
@sergei-mironov 

Check https://github.com/FrolovRuslan1/xkb-switch commits

On Wayland Plasma
```
ruslan@debian [02:27:23 PM] [~/xkb-switch] [feature/wayland_support]
-> % ./xkb-switch -d --list         
zsh: no such file or directory: ./xkb-switch
ruslan@debian [02:27:35 PM] [~/xkb-switch] [feature/wayland_support]
-> % cd build 
ruslan@debian [02:27:39 PM] [~/xkb-switch/build] [feature/wayland_support]
-> % ./xkb-switch -d --list
[DEBUG] xkb-switch version 2.1.0
/home/ruslan/xkb-switch/src/XKeyboard.cpp:156: raw layout string "us"
/home/ruslan/xkb-switch/src/XKeyboard.cpp:160: No raw variant string
num_layouts: 1
num_layouts_x11: 2
i: 0 layout: English (US)
i: 1 layout: Russian
[DEBUG] layout: us
[DEBUG] variant: <empty>
/home/ruslan/xkb-switch/src/XKeyboard.cpp:239: layout "us", variant ""
/home/ruslan/xkb-switch/src/XKeyboard.cpp:248: ... accepting with id #0
/home/ruslan/xkb-switch/src/XKeyboard.cpp:239: layout "", variant ""
us
```

**Comment by FrolovRuslan1** [comment]
@sergei-mironov @lolmaus @moceap @1nikolas @timsofteng 
Check this one, I wrote it recently
https://github.com/FrolovRuslan1/wxkb_switch
